### PR TITLE
remove boost from GUIFontCache

### PR DESCRIPTION
**Pull request by FernetMenta**
see title
superseds https://github.com/xbmc/xbmc/pull/7495

### CODE REVIEW

File: xbmc/guilib/GUIFontCache.cpp [hunk 1]
```diff
@@ -23,45 +23,66 @@
...
+      }
+    }
+
+    HashMap hashMap;
+    AgeMap ageMap;
+  };
 
   EntryList m_list;
   CGUIFontCache<Position, Value> *m_parent;
+  std::map<size_t, Value*> test;
```
Comment (afedchin) [inline xbmc/guilib/GUIFontCache.cpp]: forgotten variable :)

Comment (FernetMenta) [inline xbmc/guilib/GUIFontCache.cpp]: oops :)

**Comment by FernetMenta** [comment]
updated. this is basically the same behaviour without the boost clutter
@popcornmix you reported issues on https://github.com/xbmc/xbmc/pull/7495 could you test this on Pi please?

**Comment by Paxxi** [comment]
Will this have the same performance chracteristics as the multi_index solution? Not sure that it matters in real life use though. I'll close mine so we can focus work here.

**Comment by FernetMenta** [comment]
I can't think of any reason why it should perform worse. Maybe it is even better because index of age is simplified. Writing this I think I access this index from the wrong end :)

File: xbmc/guilib/GUIFontCache.cpp [hunk 2]
```diff
@@ -23,46 +23,66 @@
...
-    CGUIFontCacheHash<Position>, CGUIFontCacheKeysMatch < Position >
-    >
-    >
-  > EntryList;
-
-  typedef typename EntryList::template index<Age>::type::iterator EntryAgeIterator;
-  typedef typename EntryList::template index<Hash>::type::iterator EntryHashIterator;
+  struct EntryList
+  {
+    typedef std::multimap<size_t, CGUIFontCacheEntry<Position, Value>*> HashMap;
```
Comment (Paxxi) [inline xbmc/guilib/GUIFontCache.cpp]: could we get rid of all the typename stuff by switching these to `using HashMap = std::multimap<size_t, CGUIFontCacheEntry<Position, Value>*>;`? I'm a bit rusty on the rules regarding this as VS doesn't follow the standard in that regard.

**Comment by Paxxi** [comment]
hehe :) No real objections to the solution from my side but there's a few minors.
1. method names should follow naming standard with uppercase first letter
2. use std::make_pair instead of pair constructor(could make it a bit more readable)

Interested to see the testing by @popcornmix and @MilhouseVH how it performs.

File: xbmc/guilib/GUIFontCache.cpp (hunk 2 above)
Comment (FernetMenta) [inline xbmc/guilib/GUIFontCache.cpp]: I tried but had problems in defining the methods below. I am not a fan of those template stuff either. If you find way to get rid of those, be my guest.

**Comment by FernetMenta** [comment]
> 2. use std::make_pair instead of pair constructor(could make it a bit more readable)

I used value_type as suggested by the docs. This make the code shorter because you don't need the template stuff

**Comment by FernetMenta** [comment]
lets see how it compiles on the various platforms
jenkins build this please

**Comment by popcornmix** [comment]
This causes a performance regression.
On Pi2 with RSS enabled on confluence main menu with VIDEOS highlighted, I initially get ~47fps with both the original and with this PR.
However as time elapses the fps goes down with this PR. After 30 minutes I'm at 25fps. This slow down doesn't occur with original code.
I suspect cache entries are not being released correctly making cache lookups slower with time.

File: xbmc/guilib/GUIFontCache.cpp [hunk 3]
```diff
@@ -75,27 +95,26 @@ class CGUIFontCacheImpl
...
+  delete &m_key.m_text;
+  m_value.clear();
 }
 
 template<class Position, class Value>
-CGUIFontCacheEntry<Position, Value>::~CGUIFontCacheEntry()
+void CGUIFontCacheEntry<Position, Value>::Assign(const CGUIFontCacheKey<Position> &key, unsigned int nowMillis)
 {
-  delete &m_key.m_colors;
-  delete &m_key.m_text;
```
Comment (Paxxi) [inline xbmc/guilib/GUIFontCache.cpp]: should these really be removed? Assign is called on an existing entry so this could be a memory leak

Comment (FernetMenta) [inline xbmc/guilib/GUIFontCache.cpp]: I think we get fooled by github diff here. the destructor did not change and deletes are still there

**Comment by FernetMenta** [comment]
@popcornmix can you log the size of HashMap every couple of seconds?

File: xbmc/guilib/GUIFontCache.cpp (hunk 3 above)
Comment (Paxxi) [inline xbmc/guilib/GUIFontCache.cpp]: The diff does look a bit confusing but there doesn't seem to be any delete in Assign so entries being overwritten will leak the pointers

**Comment by afedchin** [comment]
@FernetMenta HashMap size isn't changed, but agemap is decreased gradually every few minutes

File: xbmc/guilib/GUIFontCache.cpp (hunk 3 above)
Comment (FernetMenta) [inline xbmc/guilib/GUIFontCache.cpp]: the old Reassign did not have any delete either

**Comment by FernetMenta** [comment]
@popcornmix @afedchin you were right. there was an issue with AgeMap.erase was called with wrong iterator

**Comment by Paxxi** [comment]
I tested my theory and it works fine on gcc which is picky about these things.
Paxxi@ba3862bb65c7be2117663413c0e07a62361ff8a9

**Comment by popcornmix** [comment]
With "squash me" commit it looks okay. Still ~47fps after 15 minutes.
Might be worth waiting for feedback from tonight's Milhouse build as this code is pretty critical for performance.

**Comment by afedchin** [comment]
@FernetMenta seems now it works as expected. both sizes of maps are identical and constant

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by Paxxi** [comment]
@FernetMenta reverted the make_pair changes Paxxi@d76d4677dbb59d762806bd7490ced0b9be942a86

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by FernetMenta** [comment]
@popcornmix @MilhouseVH did you hear any issues related to this?

**Comment by MilhouseVH** [comment]
None reported so far, but it's only been 12 hours (though 82 installations).

**Comment by popcornmix** [comment]
I'm okay with this being merged. We'll let you know if any issues show up down the line, but I think this is probably fine.

**Comment by MilhouseVH** [comment]
I'm seeing evidence of text/font "glitching" which is a result of this PR, but only with OpenELEC RPi/RPi2 builds and not OpenELEC x86 builds.

I also do not see the glitching with Kodi on Ubuntu x86, so it may be GL/GLES related.

In my RPi test builds the problem first appears in build [#0131](http://forum.kodi.tv/showthread.php?tid=250817&pid=2231580#pid2231580), which is when I introduced this PR.

In a test build I have reverted this PR on RPi/RPi2 (and also #9021, restoring Boost) and the glitching problem went away, so it's definitely a result of this PR.

The problem only seems to affect the debug overlay text, I haven't seen issues with other non-debug GUI text, but it may be a prelude to further problems down the line.

See the following frame grabs from [this video](https://www.dropbox.com/s/bp0rhf2c9tmx7vn/VID_20160206_092450.mp4?dl=1) (Dropbox, 62MB):

1. Shadow of middle line overlays top line:
screen1

2. Shadow of bottom line is much lower than it should be:
screen2

3. Foreground text (without shadow) from middle line overlays itself:
screen3

To reproduce:

1. Configure `Dim` screensaver, 20% opacity (not sure opacity matters, but it's what I'm using), timeout doesn't matter. "Black" screensaver can also be used to reproduce this issue
2. Enable debugging with debug overlay
3. Click "Preview" (or wait) for Dim/Black screensaver to activate

Result:

4. Debug overlay text starts to visibly flicker and glitch as foreground and shadow text is displayed at the wrong coordinates

Glitching of the debug overlay text continues after the screensaver is deactivated.

Using a screensaver such as Shader Toy does NOT cause this problem. I haven't tested any other screensavers.

**Pull request merged**